internal/modulecontext/module_context.go

```diff
@@ -179,6 +176,87 @@ func (m ModuleContext) BehaviorForVerb(ref schema.Ref) (optional.Option[VerbBeha
 	return optional.None[VerbBehavior](), nil
 }
 
+type ModuleContextSink = func(ctx context.Context, moduleContext ModuleContext)
+type ModuleContextSupplier interface {
+	Subscribe(ctx context.Context, moduleName string, sink ModuleContextSink)
+}
+
+type grpcModuleContextSupplier struct {
+	client ftlv1connect.VerbServiceClient
+}
+
+func NewModuleContextSupplier(client ftlv1connect.VerbServiceClient) ModuleContextSupplier {
+	return ModuleContextSupplier(grpcModuleContextSupplier{client})
+}
+
+func (g grpcModuleContextSupplier) Subscribe(ctx context.Context, moduleName string, sink ModuleContextSink) {
+	request := &ftlv1.ModuleContextRequest{Module: moduleName}
+	callback := func(_ context.Context, resp *ftlv1.ModuleContextResponse) error {
+		mc, err := FromProto(resp)
+		if err != nil {
+			return err
+		}
+		sink(ctx, mc)
+		return nil
+	}
+	go rpc.RetryStreamingServerStream(ctx, backoff.Backoff{}, request, g.client.GetModuleContext, callback)
+}
+
+// NewDynamicContext creates a new DynamicModuleContext. This operation blocks
+// until the first ModuleContext is supplied by the controller.
+//
+// The DynamicModuleContext will continually update as updated ModuleContext's
+// are streamed from the controller. This operation may time out if the first
+// module context is not supplied quickly enough (fixed at 5 seconds).
+func NewDynamicContext(ctx context.Context, supplier ModuleContextSupplier, moduleName string) (*DynamicModuleContext, error) {
+	result := &DynamicModuleContext{}
+
+	await := sync.WaitGroup{}
```

Comment:
There's a better structured concurrency package you can use, [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup). If you use `errgroup.WithContext()` and pass in a context with a deadline, it will abort the entire group, so you won't need the goroutine below, or the select.

Reply:
oh that's very nice, thanks for pointing that out